benchmarks/distributed/ddp/compare/compare_ddp.py

```diff
@@ -0,0 +1,254 @@
+            ddp_model.all_reduce_grads()
+        end.record()
+        torch.cuda.synchronize()
+        if epoch >= warmup_iterations:
+            metrics[MODEL_BACKWARD].append(start.elapsed_time(end))
+
+        # Optimization
+        optimizer.step()
+        optimizer.zero_grad()
+
+    if rank == 0:
+        print(f"\nMetrics for GPU {rank}, ddp_option={ddp_option}, buffer_size={buffer_size_in_M}M")
+        print(f"Skipped {warmup_iterations} CUDA warmpup iterations. ")
+        for step, elapsed_milliseconds in metrics.items():
+            A = np.array(elapsed_milliseconds)
+            print(' {N} iterations, {step}, mean={mean} ms, median={median} ms, p90={p90} ms, p99={p99} ms'.format(
+                N=len(A), step=step, mean=np.mean(A),
+                median=np.percentile(A, 50), p90=np.percentile(A, 90),
+                p99=np.percentile(A, 99)))
+
+        # Serialize the raw data to be used to compute summary. Didn't choose to
+        # maintain a global object holding the metrics b/c mp.spawn tries to
+        # fork all the arguments before spawning new process thus it's infeasible
+        # save global states in an object.
+        serialize(buffer_size_in_M, ddp_option, rank, metrics)
+
+def append_delta(row_list, base, exp):
+    percent = 100 * ((exp - base) / base)
+    row_list.append(percent)
+
+def print_summary(buffer_size_to_metrics):
+    # metrics: {ddp_option, Metrics}
+    # Metrics: step -> [latency]
+    for buffer_size, metrics in buffer_size_to_metrics.items():
+        assert DDPOption.DDP_CPP_CORE in metrics.keys()
+        baseline = metrics.get(DDPOption.DDP_CPP_CORE)
+        print(f"=== Summary for buffer_size: {buffer_size}M === ")
+        for step in baseline.keys():
+            # step takes value from [forward, backward]
+            # compute latency for each step into a table, each row is looks like
+            # [option, mean, diff, mean, diff, p90, diff, p95, diff, p99, diff]
+            data = []
+            baseline_latencies = baseline.get(step)
+            assert baseline_latencies is not None
+            A_baseline = np.array(baseline_latencies)
+            for ddp_option, exp_metrics in metrics.items():
+                exp_latencies = exp_metrics.get(step)
+                assert exp_latencies is not None
+                A_exp = np.array(exp_latencies)
+                # Yield option, mean, p50, p90, p95, p99 and delta.
+                row = [ddp_option]
+                row.append(np.mean(A_exp))
+                append_delta(row, np.mean(A_baseline), np.mean(A_exp))
+                for px in [50, 90, 95, 99]:
+                    base = np.percentile(A_baseline, px)
+                    exp = np.percentile(A_exp, px)
+                    row.append(exp)
+                    append_delta(row, base, exp)
+                data.append(row)
+
+            # Output buffer_size, step as a table.
+            print (tabulate(data,
```

Comment:
Shall we fix lint errors in this file as well?

Reply:
Done